### ISSUE TITLE
Simplify test database setup

### ISSUE THREAD

**Issue by kalbfled** [department-of-veterans-affairs/notification-api#1436]
## User Story - Business Need

- [ ] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As a** Notify developer,
**I want** the unit tests to create one test database, rather than one for each Pytest thread,
**So that** to simplify the code and facilitate future test improvements.

### Additional Info and Resources

As currently implemented, running unit tests results in the creation of one test database for each [Pytest thread](https://github.com/department-of-veterans-affairs/notification-api/blob/master/scripts/run_tests.sh#L28), and each database name has an [appended worker ID](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/__init__.py#L99) to name is something like "test_notification_api_gw0".

Databases have worker threads, so the more complicated multi-db setup doesn't improve performance.  Given our desire to pre-seed the test database and restore proper rollback functionality after each unit test (so tests are always repeatable), we desire a more simple, single database setup.

## Engineering Checklist

- [x] Remove all uses of [worker_id](https://pytest-xdist.readthedocs.io/en/latest/how-to.html?highlight=worker_id#identifying-the-worker-process-during-a-test).  You can find them by running `grep -rni worker_id .`.  This includes work in the VA Profile integration lambda.
- [x] After you do the above checkbox, instead of trying to fix any failing tests, please reach out to the team to inform them that tests are failing. The team will determine next steps together. (This is pointed at a 5 with this checklist item considered. Please mind the scope.) --> Everything fails; re-pointed to many more points.
- [ ] Fix all failing tests
- [ ] Delete all unused fixtures in tests/conftest.py and tests/app/conftest.py
- [ ] Delete all unused "create_" functions in tests/app/db.py

This might be all that is necessary.  The test fixture notify_db [calls create_test_db](https://github.com/department-of-veterans-affairs/notification-api/blob/master/tests/conftest.py#L85), which should not recreate a database that already exists.

## Acceptance Criteria

- [x] No significant slow down of unit tests noted
- [x] After running unit tests, the only test database in the container is "test_notification_api".

## Out of Scope

This is the first ticket in an epic.  Future tickets will:
- Restore proper rollback functionality
- Pre-seed the test database
- After the above verifications, delete the test database after it runs.  See [here](https://github.com/department-of-veterans-affairs/notification-api/blob/master/tests/conftest.py#L100).

**Comment by mjones-oddball** [comment]
@kalbfled We should remove the optional line below and just beef up this old todo ticket for that topic: https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/900
Can you add that to your todo list for next week?

> Optional: After the above verifications, consider another revision to delete the test database after it runs. See [here](https://github.com/department-of-veterans-affairs/notification-api/blob/master/tests/conftest.py#L100).

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1436?planning-poker) with Zenhub @cris-oddball @EvanParish @kalbfled @ldraney @nikolai-efimov

**Comment by kalbfled** [comment]
Having one database for all test threads causes a multitude of IntegrityError exceptions.  I suspect that this is due to tests creating instances with hard-coded IDs and other values.

**Comment by kalbfled** [comment]
The IntegrityError exceptions were occurring during test setup, and they are fairly straight-forward to eliminate by not using hard-coded values for model IDs and other unique attributes.  However, fixing the setup errors exposes the teardown errors.  Most tests use the fixture notify_db_session, which includes the teardown step of dropping tables.  This is not acceptable for a single database test setup.

I think the solution is to restore usage of the [pytest-flask-sqlalchemy](https://pypi.org/project/pytest-flask-sqlalchemy/) add-on, which we removed during the Flask upgrade last April.  The configuration and usage seems easy enough, but I will have to refactor every test that uses the notify_db or notify_db_session fixture.

**Comment by mjones-oddball** [comment]
Dave: 

> notification-api#1436 does not have a quick resolution.  Fixing the test setup errors is simple, but the teardown is a bigger issue.  The existing code drops tables, but that is inappropriate with one test database.  I'm trying to restore the rollback functionality that existing before our Flask upgrade last April.  I can spend the rest of the sprint on this.

**Comment by kalbfled** [comment]
I ran into [this issue](https://github.com/inveniosoftware/pytest-invenio/issues/100).  pytest-flask-sqlalchemy has breaks when used with flask-sqlalchemy >= 3.0.  I'm looking for a work-around.

```
request = <SubRequest '_transaction' for <Function test_get_communication_item>>, _db = <SQLAlchemy postgresql://postgres:LocalPassword@db:5432/test_notification_api +1>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7f9998b02790>

    @pytest.fixture(scope='function')
    def _transaction(request, _db, mocker):
        '''
        Create a transactional context for tests to run in.
        '''
        # Start a transaction
        connection = _db.engine.connect()
        transaction = connection.begin()
    
        # Bind a session to the transaction. The empty `binds` dict is necessary
        # when specifying a `bind` option, or else Flask-SQLAlchemy won't scope
        # the connection properly
        options = dict(bind=connection, binds={})
>      session = _db.create_scoped_session(options=options)

/usr/local/lib/python3.8/site-packages/pytest_flask_sqlalchemy/fixtures.py:38: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <SQLAlchemy postgresql://postgres:LocalPassword@db:5432/test_notification_api +1>, name = 'create_scoped_session'

    def __getattr__(self, name: str) -> t.Any:
...
>       raise AttributeError(name)
E       AttributeError: create_scoped_session

/usr/local/lib/python3.8/site-packages/flask_sqlalchemy/extension.py:1008: AttributeError
```

**Comment by kalbfled** [comment]
I have successfully run all tests in tests/app/communication_item sequentially and in parallel using a new "sample_communication_item" fixture that contains its own teardown.  I need to repeat this process for the remaining fixtures and tests.

**Comment by kalbfled** [comment]
I have refactored the heavily used sample_service, sample_user, and sample_api_key fixtures, including giving them session scope.  With these and other changes, the current state of unit tests are:
```
188 failed, 1192 passed, 147 skipped, 37 xfailed, 2084 errors
```
This is actually much better than I expected at this point.  I'm certain that many of the errors and failures have common causes, and many of the tests are parametrized.

**Comment by kalbfled** [comment]
End of day results:
```
1113 failed, 1623 passed, 147 skipped, 36 xfailed, 782 errors
```
This is after more fixtures changes.  I have some tests passing individually by failing with "detached instance" when run together sequentially.  I will push changes again when I fix that.

**Comment by kalbfled** [comment]
I have all the tests passing in tests/app/authentication.  This involved modifying core fixtures for api keys, services, and users that are used in many other tests.
```
1089 failed, 1662 passed, 150 skipped, 36 xfailed, 723 errors
```

**Comment by kalbfled** [comment]
I continue to experience expected issues during test or fixture teardown.  The problem might have something to do with application context during testing.

https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/contexts/#tests

**Comment by kalbfled** [comment]
After extensive revision to the template and notification fixtures:
```
374 failed, 1313 passed, 156 skipped, 37 xfailed, 1776 errors
```
The tests all pass in the tests/app/ folders beginning with "a" and "b", in tests/app/communication_items, and in the folders beginning with "template".  This is progress even though the total number of tests passing dropped.

**Comment by k-macmillan** [comment]
Refactored `sample_service`. It no longer calls any of the old methods that were not properly handling teardown. Order of the fixtures in a test determines teardown order! Last one first. 

Still working through services but the `test_rest.py` file (the big one in each folder) is now passing all tests.

Currently status:
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMnZSQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--cb1e62771bfd80d6894ef705c149c6e53bbb7c01/image.png)

There are some tests that absolutely fail due to mutli-core. Need to figure out how to address those. For instance some of the GET routes that gather services will fail if other workers have created. Sequentially it passes because everything cleans itself up. Single-core takes around 3 minutes to run. 5 cores (what the Github Action will use) runs in all tests in less than a minute.

My current thought is that we'll mark single-core tests as such and then run them after all the multi-core tests.

We could do it sort of like so:
```
pytest -ra -m 'not single' -n5 "$params" && pytest -ra -m "$params"
```

**Comment by kalbfled** [comment]
Results this morning with more fixes for Celery tests:
```
969 failed, 2004 passed, 349 skipped, 27 xfailed, 339 errors
```

**Comment by mjones-oddball** [comment]
Updated from an 8 to 40 because it's carried over several sprints and we do not have a sense of when it will close out.

**Comment by k-macmillan** [comment]
It was very taxing to check for the tables being cleaned up. We had to manually check each table and it was easy to miss that a table was cleaned up. Added some new methods to clean the DB by removing all the stuff that is not necessary which was setup through legacy db migrations and such. 

Created a new method that runs after any pytest command that will collect all tables that had stuff left in them. This automatically harvests all table names and queries them, reporting the ones that had more rows than expected. It is also capable of truncating all tables that have extra data so cleanup during troubleshooting is easier.

On the test-side we have finished the tests/app/ folder and made good progress through the tests/app/celery folder.

**Comment by kalbfled** [comment]
Latest run with additional Celery test fixes:
```
854 failed, 2067 passed, 449 skipped, 58 xfailed, 250 errors
```

**Comment by k-macmillan** [comment]
Merged Dave's changes and re-ran the tests:
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNlRsQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e73f01a227c108b8e5700c11561df85afeac60aa/image.png)

**Comment by kalbfled** [comment]
Latest run with additional Celery test fixes and Kyle's latest changes:
```
799 failed, 2142 passed, 449 skipped, 57 xfailed, 293 errors
```

**Comment by kalbfled** [comment]
Inching along . . .
```
793 failed, 2156 passed, 449 skipped, 57 xfailed, 294 errors
```

**Comment by k-macmillan** [comment]
Merged Dave's code and then merged master in and resolved conflicts. Current state of the tests:

```
661 failed, 2227 passed, 568 skipped, 56 xfailed, 218 errors
```

**Comment by kalbfled** [comment]
```
646 failed, 2245 passed, 568 skipped, 56 xfailed, 209 errors
```

**Comment by kalbfled** [comment]
```
636 failed, 2255 passed, 568 skipped, 56 xfailed, 208 errors
```

**Comment by mjones-oddball** [comment]
Expects to be at 100 or less tests failing end of sprint, but this will need another sprint to resolve all errors.

**Comment by kalbfled** [comment]
dao fixes:
```
578 failed, 2347 passed, 568 skipped, 56 xfailed, 166 errors
```

**Comment by k-macmillan** [comment]
Merged Dave's changes in. Updated more tests, merged master in again. Ran tests.
```
441 failed, 2515 passed, 587 skipped, 56 xfailed, 96 errors
```

**Comment by k-macmillan** [comment]
Chippin' away. Merged, updated, tested. 
```
413 failed, 2527 passed, 628 skipped, 56 xfailed, 78 errors
```

**Comment by kalbfled** [comment]
More dao test fixes:
```
387 failed, 2540 passed, 635 skipped, 56 xfailed, 83 errors
```

**Comment by kalbfled** [comment]
```
295 failed, 2638 passed, 643 skipped, 56 xfailed, 68 errors
```

**Comment by kalbfled** [comment]
```
285 failed, 2649 passed, 643 skipped, 56 xfailed, 68 errors
```

**Comment by cris-oddball** [comment]
blocked by sqlalchemy work, Kyle will move it back IP when ready.

**Comment by cris-oddball** [comment]
adding Dave's comment from Slack here:

> I blew through my remaining hours for the week trying to rebase the notification-api branch 1436-2, which is now in remote.  There are ruff (formatting, like flake8) errors.  I  fixed some of them, but more need to be fixed to run tests and see where it stands.  Nobody should merge any other branch into this branch.

**Comment by kalbfled** [comment]
I fixed all of the formatting errors and resumed fixing tests on the 1436-2 branch.  There are some regressions, relative to where this stood before the holidays, but I think they will be easy to fix.

```
349 failed, 2636 passed, 640 skipped, 55 xfailed, 68 errors
```

The tests in these subfolders as tests/ pass with proper database clean-up:
- lambda_functions/
- app/a* (all the 'a' folders)
- app/billing/
- all the app/c* folders other than app/celery/

All tests pass in app/delivery/, but the teardown is not proper (leave rows in databases).  I will resume in the morning by assessing additional folders.  Most of the remaining work is in app/celery/ and app/dao/.

**Comment by kalbfled** [comment]
```
340 failed, 2646 passed, 640 skipped, 55 xfailed, 64 errors
```

These are the outstanding failures:
- tests/app/celery/ 308 pass of 419
- tests/app/dao/ 505 pass of 767
- tests/app/delivery/ All tests pass, but the cleanup leaves artifacts.  Ed is working on this.
- tests/app/notifications/ 267 pass of 431

**Comment by kalbfled** [comment]
```
268 failed, 2723 passed, 640 skipped, 55 xfailed, 53 errors
```

The improper cleanup in tests/app/delivery/ is resolved.

**Comment by kalbfled** [comment]
```
241 failed, 2750 passed, 640 skipped, 55 xfailed, 46 errors
```
All tests in tests/app/notifications/ pass with run as a subset.  When running the full suite, 3 of those tests fail (presumably because of teardown issues in other tests).

**Comment by kalbfled** [comment]
```
227 failed, 2764 passed, 640 skipped, 56 xfailed, 45 errors
```

**Comment by k-macmillan** [comment]
Getting closer!
```
197 failed, 2793 passed, 642 skipped, 56 xfailed, 36 errors
```

**Comment by kalbfled** [comment]
I ran Kyle's latest changes with 11 threads and got these results:
```
232 failed, 2757 passed, 642 skipped, 56 xfailed, 41 errors
```
Then I ran my latest changes on top of Kyle's changes and got these results:
```
196 failed, 2793 passed, 642 skipped, 56 xfailed, 34 errors
```
Based on this, I pushed my latest changes, which primarily fix the tests in tests/app/dao/test_ft_billing_dao.py.

**Comment by mchlwellman** [comment]
I pulled Ed's changes, and with my own these are the results now - 

 `179 failed, 2809 passed, 645 skipped, 56 xfailed, 34 errors `

**Comment by edDocMe360** [comment]
163 failed, 2826 passed, 645 skipped, 56 xfailed, 28 errors in 697.11s (0:11:37) =

**Comment by k-macmillan** [comment]
Pulled, merged, tested with my changes. Seeing the following with single-core:
```
 86 failed, 2901 passed, 650 skipped, 56 xfailed, 20 errors
```
Pushed.

**Comment by kalbfled** [comment]
Running 11 threads with Kyle changes plus more of my fixes:
```
137 failed, 2849 passed, 650 skipped, 56 xfailed, 23 errors
```

**Comment by edDocMe360** [comment]
= 129 failed, 2857 passed, 650 skipped, 56 xfailed, 24 errors in 912.32s

**Comment by kalbfled** [comment]
```
108 failed, 2879 passed, 650 skipped, 56 xfailed, 20 errors
```

**Comment by mchlwellman** [comment]
`102 failed, 2885 passed, 650 skipped, 56 xfailed, 21 errors`

**Comment by kalbfled** [comment]
11 threads
```
95 failed, 2892 passed, 650 skipped, 56 xfailed, 22 errors
```

**Comment by mchlwellman** [comment]
1 Thread:

`41 failed, 2949 passed, 650 skipped, 56 xfailed, 16 errors`

11 Threads:
`103 failed, 2885 passed, 650 skipped, 56 xfailed, 26 errors`

I'm not sure why my numbers are higher, I am up to date with github.

**Comment by kalbfled** [comment]
11 threads
```
87 failed, 2900 passed, 650 skipped, 56 xfailed, 20 errors
```
1 thread
```
34 failed, 2954 passed, 650 skipped, 56 xfailed, 14 errors
```
This reflects primarily teardown fixes.

**Comment by kalbfled** [comment]
I fixed tests/app/dao/test_login_event.py.  With 11 threads:
```
84 failed, 2903 passed, 650 skipped, 56 xfailed, 16 errors
```

**Comment by k-macmillan** [comment]
Single worker:
```
23 failed, 2957 passed, 660 skipped, 55 xfailed, 8 errors
```

**Comment by kalbfled** [comment]
11 threads
```
72 failed, 2908 passed, 659 skipped, 55 xfailed, 16 errors
```

**Comment by k-macmillan** [comment]
Currently failing files:
- tests/app/celery/test_process_ses_receipts_tasks.py
- tests/app/celery/test_provider_tasks.py
- tests/app/dao/test_services_dao.py
- tests/app/dao/test_templates_dao.py
- tests/app/dao/test_users_dao.py
- tests/app/service/test_rest.py

**Comment by k-macmillan** [comment]
## Multi-worker results
```
91 failed, 2888 passed, 660 skipped, 55 xfailed, 14 errors
```


## Single worker results
```
11 failed, 2969 passed, 660 skipped, 55 xfailed, 5 errors
```

Currently failing files:
- tests/app/dao/test_services_dao.py
- tests/app/dao/test_templates_dao.py
- tests/app/dao/test_users_dao.py
- tests/app/service/test_rest.py

The current failures are almost certainly related to teardowns.
Full status:
```
FAILED tests/app/dao/test_services_dao.py::test_get_all_services - assert 3 == 1
FAILED tests/app/dao/test_services_dao.py::test_get_all_services_should_return_in_created_order - assert 6 == 4
FAILED tests/app/dao/test_services_dao.py::test_get_all_services_should_return_empty_list_if_no_services - assert 2 == 0
FAILED tests/app/dao/test_services_dao.py::test_create_service_and_history_is_transactional - assert [<Service 2ef...6f4fa6b8af91>] == []
FAILED tests/app/dao/test_services_dao.py::test_delete_service_and_associated_objects - assert [(<Template 1...1635773015>,)] == []
FAILED tests/app/dao/test_templates_dao.py::test_create_only_one_template[sms] - assert 3 == 1
FAILED tests/app/dao/test_templates_dao.py::test_create_only_one_template[email] - assert 4 == 1
FAILED tests/app/dao/test_users_dao.py::test_create_only_one_user - assert 7 == 1
FAILED tests/app/dao/test_users_dao.py::test_get_all_users - assert 9 == 2
FAILED tests/app/service/test_rest.py::test_get_service_list_should_return_empty_list_if_no_services - AssertionError: assert 12 == 0
FAILED tests/app/service/test_rest.py::test_get_services_with_detailed_flag - AssertionError: assert 13 == 1
ERROR tests/app/dao/test_jobs_dao.py::test_set_scheduled_jobs_to_pending_ignores_jobs_scheduled_in_the_future - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_set_scheduled_jobs_to_pending_ignores_jobs_scheduled_in_the_future> (9 sub-exceptions)
ERROR tests/app/dao/test_jobs_dao.py::test_get_future_scheduled_job_gets_a_job_yet_to_send - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_get_future_scheduled_job_gets_a_job_yet_to_send> (9 sub-exceptions)
ERROR tests/app/dao/test_provider_rates_dao.py::test_create_provider_rates - sqlalchemy.exc.NoResultFound: No row was found when one was required
ERROR tests/app/dao/test_templates_dao.py::test_create_only_one_template[sms] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_create_only_one_template[sms]> (6 sub-exceptions)
ERROR tests/app/dao/test_templates_dao.py::test_create_only_one_template[email] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_create_only_one_template[email]> (6 sub-exceptions)
```

**Comment by kalbfled** [comment]
11 threads
```
64 failed, 2916 passed, 660 skipped, 55 xfailed, 8 errors
```

**Comment by kalbfled** [comment]
11 threads
```
56 failed, 2924 passed, 660 skipped, 55 xfailed, 9 errors
```

**Comment by kalbfled** [comment]
All tests pass when run sequentially!  With 11 threads:
```
65 failed, 2915 passed, 660 skipped, 55 xfailed, 11 errors
```
There are artifacts in these tables: 
- daily_sorted_letter
- va_profile_local_cache

The latter is expected and only affects one suite of lambda function unit tests.

**Comment by kalbfled** [comment]
After marking additional tests as "serial", the status with 9 threads is:
```
35 failed, 2875 passed, 660 skipped, 55 xfailed, 3 errors
```
All tests continue to pass when run singled-threaded.  I am in the process of evaluating the remaining failures and either fixing them or marking them a "serial" as well.  Kyle and I discussed a new ticket for revising "serial" tests for parallel.

**Comment by kalbfled** [comment]
With 9 threads in parallel + 75 tests run sequentially:
```
28 failed, 2953 passed, 660 skipped, 55 xfailed, 7 errors
```
I added the `-q` flag to run_tests.sh to lessen the output while tests run.

**Comment by kalbfled** [comment]
With 9 threads in parallel + 80 tests run sequentially:
```
24 failed, 2954 passed, 660 skipped, 55 xfailed, 4 errors
```

**Comment by kalbfled** [comment]
With 9 threads in parallel + 80 tests run sequentially:
```
19 failed, 2958 passed, 660 skipped, 55 xfailed, 4 errors
```

**Comment by kalbfled** [comment]
With 9 threads in parallel:
```
4 failed, 2896 passed, 660 skipped, 55 xfailed, 1 error
```
With 107 tests running sequentially:
```
4 failed, 103 passed, 3589 deselected
```

Serial failures:
```
FAILED tests/app/dao/test_provider_details_dao.py::test_can_get_sms_providers_in_order_of_priority
FAILED tests/app/dao/test_provider_details_dao.py::test_get_current_sms_provider_returns_provider_highest_priority_active_provider
FAILED tests/app/dao/test_provider_details_dao.py::test_switch_sms_provider_to_current_provider_does_not_switch
FAILED tests/app/dao/test_provider_details_dao.py::test_switch_sms_provider_to_inactive_provider_does_not_switch
```

**Comment by kalbfled** [comment]
I just got a clean parallel run, but these tests fail when running sequentially:
```
FAILED tests/app/dao/test_provider_details_dao.py::test_can_get_email_providers_in_order_of_priority
FAILED tests/app/dao/test_provider_details_dao.py::test_update_adds_history
FAILED tests/app/dao/test_provider_details_dao.py::test_updated_at - sqlalche...
FAILED tests/app/dao/test_provider_details_dao.py::test_dao_get_provider_stats_returns_data_in_type_and_identifier_order
FAILED tests/app/delivery/test_send_to_providers.py::test_should_return_highest_priority_active_provider
========== 5 failed, 137 passed, 3554 deselected in 121.54s
```

**Comment by kalbfled** [comment]
All tests pass or are skipped/x-failed.  Given the long running nature of this ticket, @k-macmillan and I decided to create new tickets to clean up residual tech debt, etc.  See #1631 and XXXX.
